src/foam/nanos/auth/ruler/AddPasswordHistoryAction.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.auth.ruler',
+  name: 'AddPasswordHistoryAction',
+  documentation: 'Adds an entry into the password history of a user',
+
+  implements: ['foam.nanos.ruler.RuleAction'],
+
+  javaImports: [
+    'foam.nanos.logger.Logger',
+    'foam.nanos.auth.HistoricPassword',
+    'foam.nanos.auth.User',
+    'java.util.Date'
+  ],
+
+  constants: [
+    {
+      name: 'MAXIMUM_HISTORIC_PASSWORDS',
+      type: 'Long',
+      value: 10
+    }
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        User user = (User) obj;
```

Comment:
```suggestion
        User user = (User) oldObj;
```

You're using the new user object here instead of the old one, and since this rule only applies on update (not create), it would be adding the new password to the array instead of the old one. Alternatively, you could change the operation of this rule to apply on updates and creates and then I think it would be correct.